### ISSUE TITLE
__getitem__ is not allowed for functions nor methods

### ISSUE THREAD

**Issue by wasilukm** [PyCQA/pylint#1627]
I have a library that provides \_\_getitem\_\_ interface for some methods, therefore user code may call  a method with index, for example:
some_obj.some_method\[1\]\(\)

Example code which reproduces this:
```python
def getitem_provider(wrapped):
    class Wrapper(object):
        def __getitem__(self, index):
            print 'index', index
            return wrapped
    return Wrapper()


@getitem_provider
def foo():
    pass

foo[1]()
```

### Current behavior
Error unsubscriptable-object is reported by Pylint for line with ```foo[1]()``` even if the code is working fine.

### Expected behavior
I would expect that Pylint correctly checks if \_\_getitem\_\_ is available for functions/methods or there is a selective suppressor that allows disabling unsubscriptable-object error for specific modules or classes.

### pylint --version output
No config file found, using default configuration
pylint 1.7.2,
astroid 1.5.3
Python 2.7.12 (v2.7.12:d33e0cf91556, Jun 27 2016, 15:19:22) [MSC v.1500 32 bit (Intel)]

**Comment by rogalski** [comment]
We currently do not understand decorators in this way and we assume that decorators does not change public API of decorated object.

**Comment by Pierre-Sassoulas** [comment]
This has been fixed in pylint 2.15, I can't reproduce with:
```
def getitem_provider(wrapped):
    class Wrapper:
        def __getitem__(self, index):
            print("index", index)
            return wrapped

    return Wrapper()


@getitem_provider
def example():
    pass


example[1]()
```
